ui/packages/consul-peerings/app/templates/dc/peers/index.hbs

```diff
@@ -83,62 +136,97 @@ as |sort filters items|}}
           />
         </BlockSlot>
         <BlockSlot @name="content">
-          <DataCollection
-            @type="peer"
-            @sort={{sort.value}}
-            @filters={{filters}}
-            @search={{search}}
-            @items={{items}}
-          as |collection|>
-            <collection.Collection>
+
+            <ModalDialog
+              @aria={{hash
+                label="Regenerate token"
+              }}
+              @onclose={{set this 'item' undefined}}
+            as |modal|>
+              <BlockSlot @name="header">
+                {{did-insert (set this 'regenerate' modal)}}
```

Comment:
this is a pattern we really should look into - ideally this would just be yielded by the modal-dialogue itself to be able to do that

can we just yield out the action to open/close the modal [here](https://github.com/hashicorp/consul/blob/main/ui/packages/consul-ui/app/components/modal-dialog/index.hbs#L5) and then not use this register pattern?

Reply:
Not totally sure I understand, the action to open/close the modal is being yielded by the modal-dialog itself. In the above code the `|modal|` is the yielded API:

```hbs
<ModalDialog ... as |modal|>
```

It gets yielded from inside the component [here](https://github.com/hashicorp/consul/blob/8d4baafd84532cb950a810fe494e16c4d9a2e559/ui/packages/consul-ui/app/components/modal-dialog/index.hbs#L40-L45)

In case it helps this is roughly the equivalent to the following javascript:

```javascript
class Template {
  constructor() {
    modal({aria: {label: 'Regenerate token'}}, (modal) => this.regenerate = modal);
    // use the modals API
    this.regenerate.close()
  }
}
```

Ideally it would be more like:

```javascript
class Template {
  constructor() {
    this.regenerate = modal({aria: {label: 'Regenerate token'}});
    // use the modals API
    this.regenerate.close()
  }
}
```

but that's a consequence of embers component structure.

Could you help me to understand your query further as I maybe don't follow?

Reply:
My question points to the following:

You are using a a render helper here to set the `regenerate` property on the template-context (the controller) here to the object that you are yielding out in the header block-slot. As far as I can tell this is basically what is happening here:

```js
this.set('regenerate', modalApiYieldedToHeaderBlockSlot)
```

You seem to need to do that because you want to interact with the modal from [other places of the template](https://github.com/hashicorp/consul/blob/8e27be16003215cd41a9f64fa6b7ca79852aafbd/ui/packages/consul-peerings/app/templates/dc/peers/index.hbs#L110). This is pretty unusual. Ember provides the ability to `yield` for being able to use the yielded values directly:

```hbs
<Modal as |modal|>
  <button type="button" {{on "click" modal.open}}>
    Open modal
  </button>
    <button type="button" {{on "click" modal.close}}>
    Close modal
  </button>
</Modal>
```

Instead of accessing the yielded api directly we are registering the yielded object to the template context. The way we are using the block slots seems to get into the way here - can we yield the modal api (maybe via different slot) so the above would be possible? Or is this a bigger refactor? Would this work?

```hbs
{{!-- ModalDialog.hbs --}}
<YieldSlot @name="api" @params={{block-params (hash
  open=(action "open")
  close=(action "close")
  opened=this.isOpen
  aria=aria
)}}>
  {{yield}}
</YieldSlot>

{{!-- ... Rest of modal template --}}

{{!-- and then use it like this --}}

<ModalDialog>
  {{!-- Rest of template ... --}}
   <BlockSlot @name="api" as |modal|>
    <Action
      data-test-create
      class="type-create"
      {{on "click" modal.open}}
     >
       Add peer connection
     </Action>
  </BlockSlot>
</ModalDialog>
```
The way the modal-dialogue uses the yield-slots doesn't seem to be the way they work. you have to pass the `params` instead of yielding to its content block apparently.

Using it like this would mean we won't need to register an object on the template context to be able to interact with the modal-component with the additional benefit that everything gets cleaned up properly when the template unrenders and we don't set properties that we then don't unset on the template context anymore.

Reply:
@LevelbossMike @johncowen This seems like a worthwhile discussion to have/revisit in the near future as we are able to dedicate more time to examining this implementation in detail, and determining potential paths for a refactor.

For now with regards to this modal implementation, given that this pattern is what we're using for other modals in the app, I would advise that we stick with what we have and focus on whether there are any major issues with this particular instance of the implementation. Let's make sure to log this as an item to come back around and do a technical review on in a few weeks post 1.13.

Reply:
The modal component is one of the components I've been wanting to upgrade for a very long time, but not been given the opportunity to do so. For historical background, a while ago I slowly but surely converted the majority of our components that use the Ember 2.18 supported block slots to Ember 3 supported slots (or contextual components) but I had to stop part way through - our Modal component is one of the ones that is still left over. So yes, lets upgrade it, but just not right now.

One thing I've noticed here which might help with some understanding:

```javascript
<ModalDialog>
  {{!-- Rest of template ... --}}
   <BlockSlot @name="api" as |modal|>
    <Action
      data-test-create
      class="type-create"
      {{on "click" modal.open}}
     >
       Add peer connection
     </Action>
  </BlockSlot>
</ModalDialog>
```

If the button to open the modal is inside the modal panel itself, how then do you click that button when the modal is closed? Just to be clear the component here is the modal panel itself, like the overlay and 'window' for the modal.